### PR TITLE
working demo with tests for toMermaid()

### PR DESCRIPTION
**Pull request by khill-fbmc**
image

The PR is here!

So I changed:
 - README has strings for example instead of integers on the enums
 - Extracted the `Door` class in the tests to make it re-useable
 - Saw there was a duplicate transition in the door class and removed it
 - Brought some dependencies up to more recent versions

Added:
 - StateMachine#toMermaid() method to output mermaid js code
 - Tests to cover the output of the `Door` class's `toMermaid()`
 - `tools` dir with a script to write-out the mermaid code of the `Door` class
 - package.json script to run the generator `npm run diagram`

I'm happy to change things if needed :)

### CODE REVIEW

**Comment by khill-fbmc** [comment]
This addresses #22

File: README.md [hunk 1]
```diff
@@ -19,18 +19,18 @@ I'm using this state-machine as a simple replacement for Redux in some ReactJs b
 
 ## Get it
 
-  ```script
```
Comment (bennycode) [inline README.md:22]: There are a lot of formatting changes. Can you revert these to make the PR more directed towards your changes?

Comment (khill-fbmc) [reply README.md:22]: Whoops! I sometimes forget I have `formatOnSave` turned on and not everyone wants the same format as me :)

I'll redo the pull when I get a chance.

File: package.json [hunk 2]
```diff
@@ -29,21 +29,22 @@
...
   "devDependencies": {
-    "@types/jest": "^29.0.0",
-    "@typescript-eslint/eslint-plugin": "^6.12.0",
-    "@typescript-eslint/parser": "^6.12.0",
-    "eslint": "^8.54.0",
-    "eslint-config-airbnb-typescript": "^17.1.0",
+    "@types/jest": "^29.5.14",
+    "@typescript-eslint/eslint-plugin": "^7.18.0",
+    "@typescript-eslint/parser": "^7.18.0",
+    "eslint": "^8.57.1",
```
[the remark this answers is not in the document]
Comment (khill-fbmc) [reply package.json:39]: No. I just have a compulsion to update things when they are "old" I often get carried away from my intended changes.

File: .eslintrc.json [hunk 3]
```diff
@@ -20,83 +19,105 @@
   "rules": {
     "import/no-extraneous-dependencies": "off",
     "import/extensions": "off",
-    "no-null/no-null": "error",
-    "sonarjs/cognitive-complexity": ["error", 30],
+    "sonarjs/cognitive-complexity": [
```
Comment (eram) [inline .eslintrc.json:22]: why are we making changes to this file?

File: README.md [hunk 4]
```diff
@@ -44,72 +44,85 @@ Let's code it in Typescript! Note that the same code can be run in Javascript, j
 import { t, StateMachine } from "typescript-fsm";
 
 // these are the states and events for the door
-enum States { closing = 0, closed, opening, opened, broken };
-enum Events { open = 100, openComplete, close, closeComplete, break };
+enum States {
```
Comment (eram) [inline README.md:47]: pls revert

File: src/stateMachine.ts [hunk 5]
```diff
@@ -49,20 +49,31 @@ export class StateMachine<
     });
   }
 
+  /**
+   * Get the current state of the machine.
+   */
   getState(): STATE { return this._current; }
 
+  /**
+   * Test if a given even can transition to a new state.
```
Comment (eram) [inline src/stateMachine.ts:58]: event

File: tools/GenerateDiagram.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { writeFileSync } from "node:fs";
```
Comment (eram) [inline tools/GenerateDiagram.ts:1]: remove file

File: tsconfig.json [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {
   "compilerOptions": {
+    // "strict": true,
```
Comment (eram) [inline tsconfig.json:3]: pls revert all file changes

**Review by eram — CHANGES_REQUESTED** [review]
I was expecting only a new function and a test, but I've got a whole bunch of fixes.

**Comment by khill-fbmc** [comment]
I apologize. My messy style of work in my own git repos is spilling. I didn't mean to touch so many things.

I will redo the whole PR :)

File: .eslintrc.json (hunk 3 above)
Comment (khill-fbmc) [reply .eslintrc.json:22]: I was thinking that we could use `strict` mode in typescript, but that made different type errors appear. Again, apologies for scattering things about.

**Comment by khill-fbmc** [comment]
> I was expecting only a new function and a test, but I've got a whole bunch of fixes.

I have never really worked on other peoples projects or on a team. I've always been a solo dev, so I have some bad habits.

Please don't look in my repos 😆

**Pull request closed without merging**